proxyserver/middleware/versioned_writes.go

```diff
@@ -0,0 +1,549 @@
+	}
+}
+
+func (l *listingWriter) Write(stuff []byte) (int, error) {
+	if l.success {
+		l.buffer.Write(stuff)
+	}
+	return len(stuff), nil
+}
+
+type VersionedObjectWriter struct {
+	header http.Header
+	status int
+}
+
+func (vcw *VersionedObjectWriter) Write(stuff []byte) (int, error) {
+	return len(stuff), nil
+}
+
+func (vcw *VersionedObjectWriter) Header() http.Header {
+	return vcw.header
+}
+
+func (vcw *VersionedObjectWriter) WriteHeader(status int) {
+	vcw.status = status
+}
+
+func NewVersionedObjectWriter() *VersionedObjectWriter {
+	header := make(map[string][]string)
+	return &VersionedObjectWriter{
+		status: 500,
+		header: header,
+	}
+}
+
+type VersionedContainerWriter struct {
+	http.ResponseWriter
+}
+
+func (vcw *VersionedContainerWriter) WriteHeader(status int) {
+	mode := ""
+	location := ""
+	if vcw.ResponseWriter.Header().Get(SYSMETA_VERSIONS_LOC) != "" {
+		location = vcw.ResponseWriter.Header().Get(SYSMETA_VERSIONS_LOC)
+	}
+	if vcw.ResponseWriter.Header().Get(SYSMETA_VERSIONS_MODE) != "" {
+		mode = vcw.ResponseWriter.Header().Get(SYSMETA_VERSIONS_MODE)
+	}
+
+	if location != "" {
+		if mode == "history" {
+			vcw.ResponseWriter.Header().Set(CLIENT_HISTORY_LOC, location)
+		} else {
+			vcw.ResponseWriter.Header().Set(CLIENT_VERSIONS_LOC, location)
+		}
+	}
+
+	vcw.ResponseWriter.WriteHeader(status)
+}
+
+func (v *versionedWrites) handleContainer(writer http.ResponseWriter, request *http.Request) {
+	versionsLocs, v_ok := request.Header[CLIENT_VERSIONS_LOC]
+	historyLocs, h_ok := request.Header[CLIENT_HISTORY_LOC]
+
+	if v_ok && h_ok {
+		srv.SimpleErrorResponse(writer, 400, fmt.Sprintf("Only one of %s or %s may be specified", CLIENT_VERSIONS_LOC, CLIENT_HISTORY_LOC))
+		return
+	}
+
+	if v_ok || h_ok {
+		val := ""
+		var mode string
+		if v_ok {
+			val = versionsLocs[0]
+			mode = "stack"
+		} else {
+			val = historyLocs[0]
+			mode = "history"
+		}
+		if val != "" {
+			request.Header.Set(SYSMETA_VERSIONS_LOC, val)
+			request.Header.Set(SYSMETA_VERSIONS_MODE, mode)
+
+			request.Header.Del(CLIENT_VERSIONS_LOC)
+			request.Header.Del(CLIENT_HISTORY_LOC)
+			// Adding a location trumps removing them.
+			request.Header.Del("X-Remove-Versions-Location")
+			request.Header.Del("X-Remove-History-Location")
+		} else {
+			request.Header.Set("X-Remove-Versions-Location", "x")
+		}
+	}
+
+	if request.Header.Get("X-Remove-Versions-Location") != "" || request.Header.Get("X-Remove-History-Location") != "" {
+		request.Header.Set(CLIENT_VERSIONS_LOC, "")
+		request.Header.Set(SYSMETA_VERSIONS_LOC, "")
+		request.Header.Set(SYSMETA_VERSIONS_MODE, "")
+		request.Header.Del("X-Remove-Versions-Location")
+		request.Header.Del("X-Remove-History-Location")
+	}
+
+	vcw := &VersionedContainerWriter{
+		ResponseWriter: writer,
+	}
+
+	v.next.ServeHTTP(vcw, request)
+}
+
+func (v *versionedWrites) versionedObjectPrefix(object string) string {
+	return fmt.Sprintf("%03x/%s", len(object), object)
```

Comment:
swift's looks like:
003cat/1500059950.32742
this looks like:
003/cat1500060065.79092

we need to match swift i think so people can upgrade. as it is if you put a bunch of versioned stuff in swift, switch to hbird and do a delete, it doesn't restore prev copies